src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThread.java

```diff
@@ -97,7 +97,7 @@
     final FlowHead head;
 
     @Nullable
-    private final ContextVariableSet contextVariables;
+    private ContextVariableSet contextVariables;
```

Comment:
Maybe this should be volatile? Synchronization is taken care of in `CpsBodyExecution`s handlers, but I think `getContextVariables` could be called from other threads so we need to make sure they see the latest value for reads. I need to spend a bit more time thinking about this.

Reply:
> `getContextVariables` could be called from other threads

Whatever calls `StepContext.get(Class)`, yes. But all that should be done by the time the body completes anyway.